platform/engineering/design-docs/micro-frontend-architecture.md

```diff
@@ -0,0 +1,255 @@
+
+#### The State of `vets-website`
+
+##### Build
+The `vets-website` build is responsible for a lot of things. The most important
+pieces include:
+- Pulling content from the [CMS](#CMS)
+- Building that content into HTML pages
+- Validating the content for broken links, accessibility violations, etc.
+- Building all JavaScript bundles
+  - Both the common bundle and individual application bundles
+- Building all CSS bundles
+  - Both common and application-specific
+- Running all >5k unit tests for all applications
+- Running all end-to-end tests for all applications
+
+
+##### Deploy
+When `vets-website` is deployed, all content and applications are deployed
+together. **There is no way to deploy one application without deploying
+everything.**
+
+There is, however, a way to deploy the latest content from the CMS without
+deploying the latest application code. This is called a **content-only deploy**.
+The latest `vets-website` release is used instead of building code in the latest
+commit of the `vets-website` `master` branch.
+
+
+#### The Problems
+##### Failures in one Service affect all other Services
+When one application team introduces test failures into `master`, all PRs from
+all teams that include the offending commit in `master` will fail in Continuous
+Integration (CI) because all tests are run, not just the tests for a single
+application.
+
+**Example:** Team A introduces a test failure in `master`. Team B has an urgent
+bug fix that needs deployed as soon as possible, so they make a branch of
+`master` and quickly fix it. Team B's branch can't be merged until Team A's
+unrelated test failure is fixed.
+
+Likewise, when there's a content issue, such as an accessibility violation, all
+application teams' PRs will fail in CI until the accessibility violation has
+been resolved by a content editor.
+
+**Important exception:** Test failures in the latest `master` branch will not prevent
+content-only deployments since those use the latest stable `vets-website`
+release.
+
+
+###### Test failures in `master`? How?
+This happens most frequently when two individual branches work on similar parts
+of the code. The tests on each branch fail, but the combined changes cause the
+result—`master`—to fail.
+
+**TODO:** Use an actual example of where this happened.
+
+
+##### Developing applications requires CMS content
+To run `vets-website` locally for application development, front end engineers
+must either have access to the CMS content directly via the SOCKS proxy or fetch
+the latest stable cached content from an AWS S3 bucket. Not having this content
+will result in a failure at build time.
+
+<!-- What other problems are we running into with the current architecture? -->
+<!-- Code ownership? -->
+<!-- Dependency management? -->
+<!-- Deploy schedule? -->
+
+
+### High Level Design
+
+**Each application** will be version controlled in its own **separate
+repository** with its own **CI pipeline** and **deployment schedule**.
```

Comment:
I wanted to dig up an old documented decision on why we went with a monorepo, but couldn't find it...

Does using something like `lerna` within a monorepo with each app being its own package work within your vision?

Reply:
@U-DON I'd definitely be interested to hear more on the monorepo vs. separate repos tradeoffs. I think Dror will as well once he arrives (already had a chat with him about it).

@cvalarida is there a way we can get some/many of the MFE benefits _without_ separating out repos? that seems like a line of research at least worth evaluating the pros/cons of

Reply:
Yup! This is definitely something I thought about. I think something like `lerna` is _interesting_, but I don't know what advantages it would have. My understanding is fairly limited. From what I can tell, the advantages of a monorepo are minimal, and it's yet more overhead engineers need to understand before they can be productive.

Mostly I was concerned the CI. Specifically, if I'm making a PR to App 1, how do we tell the CI to only run unit tests on App 1? (Because the rest of the unit tests don't apply.) It's probably possible, but I have no idea how to do it.

Additionally, I'm concerned about incidental coupling between applications. Specifically, it'd be easy for engineers working on App 1 to import something from App 2 without the App 2 team knowing about it. When the App 2 team makes a change to that thing, it'll affect App 1. If the repos were split up, this would still be possible, I think, but it would be harder. Like, you'd have to deliberately list App 2 as a dependency in `package.json` rather than just tossing an `import` statement in a JS file.

That said, I'm not very opinionated on this. I like the clear separation of concerns we get when splitting out the applications into their own repositories, but I'm definitely not against the monorepo approach. I'm super interested in hearing what other folks have to say.

Reply:
as a reference point, part of how Google's monorepo succeeds is a CI system that tries to determine "affected tests" for a given change, and only run those. it generally depends on [Bazel](https://bazel.build/) to do that, by using the dependency tree of build targets.

that said, some of the concepts are still usable. the naive approach I'd start with: if code and tests are located in structurally consistent ways, we could probably provide utilities to run only tests within certain directories (test filters?), and then whatever script the CI executes decides based on which files are modified which tests need to get run. changes to common code probably still run all tests. is that something that'd be workable in our repo?

Reply:
^ This is actually an approach I took to running the unit tests for the [early version of a CircleCI workflow](https://github.com/department-of-veterans-affairs/vets-website/blob/a969e53de3fa7066119d9fc9dc4a6c9914642bb3/script/run-isolated-tests.sh) I adapted from our current Jenkins pipeline. It's a check that's running on PRs now (non-blocking), so we can observe its efficacy.

Reply:
- I don't think Lerna really meets our use case. Don't think of it as a "multiple package in one repo" tool but as a "tool to manage multiple npm modules published from a single repo." It runs npm scripts from the root directory- running package specific scripts can be a challenge with the way it manages dependencies. 
  - We should evaluate it- I'm not an expert. But I had issues with things like running package specific linting rules from the root npm scripts. I don't think its behavior is intuitive and could be confusing to VFS teams. 
- I would like to encourage better boundaries and decoupling of the website pieces. Some of this has led to issues like app specific concerns in the global build script, forms system, and helpers.
  - separating the forms system from the website repo was problematic because of how entangled it is with apps and with how poorly it separates concerns 
- Versioning apps can encourage behaviors that reduce friction during ownership changes (app can change ownership every 1-2 year): 
  - documenting changes in release notes 
  - thinking of changes in terms of explicit releases